src/agent/coverage/src/block/pe_provider.rs

```diff
@@ -0,0 +1,623 @@
+
+        while let Some(module) = modules.next()? {
+            if let Some(module_info) = self.pdb.module_info(&module)? {
+                let mut symbols = module_info.symbols()?;
+                while let Some(symbol) = symbols.next()? {
+                    if let Ok(SymbolData::Procedure(proc)) = symbol.parse() {
+                        visitor.visit_procedure_symbol(&proc)?;
+                    }
+                }
+            }
+        }
+
+        let global_symbols = self.pdb.global_symbols()?;
+        let mut iter = global_symbols.iter();
+
+        while let Some(symbol) = iter.next()? {
+            if let Ok(SymbolData::Procedure(proc)) = symbol.parse() {
+                visitor.visit_procedure_symbol(&proc)?;
+            }
+        }
+
+        Ok(visitor.access_offsets)
+    }
+
+    fn provide_from_pcs_table(&mut self, pcs_table: SancovTable) -> Result<BTreeSet<u32>> {
+        // Read the PE directly to extract the PCs from the PC table.
+        let pe_alignment = self
+            .pe
+            .header
+            .optional_header
+            .ok_or_else(|| format_err!("PE file missing optional header"))?
+            .windows_fields
+            .file_alignment;
+        let pe_offset = goblin::pe::utils::find_offset(
+            pcs_table.offset as usize,
+            &self.pe.sections,
+            pe_alignment,
+        );
+        let pe_offset =
+            pe_offset.ok_or_else(|| format_err!("could not find file offset for sancov table"))?;
+        let table_range = pe_offset..(pe_offset + pcs_table.size);
+        let pcs_table_data = self
+            .data
+            .get(table_range)
+            .ok_or_else(|| format_err!("sancov table slice out of file range"))?;
+
+        if pcs_table_data.len() % 16 != 0 {
+            anyhow::bail!("invalid PC table size");
+        }
+
+        let mut pcs = BTreeSet::default();
+
+        let module_base: u64 = self.pe.image_base.try_into()?;
+
+        // Each entry is a struct with 2 `uintptr_t` values: a PC, then a flag.
+        // We only want the PC, so start at 0 (the default) and step by 2 to
+        // skip the flags.
+        for chunk in pcs_table_data.chunks(8).step_by(2) {
+            let le: [u8; 8] = chunk.try_into()?;
+            let pc = u64::from_le_bytes(le);
+            let pc_offset: u32 = pc
+                .checked_sub(module_base)
+                .ok_or_else(|| {
+                    format_err!(
+                        "underflow when computing offset from VA: {:x} - {:x}",
+                        pc,
+                        module_base,
+                    )
+                })?
+                .try_into()?;
+            pcs.insert(pc_offset);
+        }
+
+        Ok(pcs)
+    }
+}
+
+/// Searches a PDB for data symbols that delimit various Sancov tables.
+#[derive(Default)]
+pub struct SancovDelimiterVisitor<'am> {
+    address_map: AddressMap<'am>,
+
+    llvm_bools_start: Option<u32>,
```

Comment:
Since these always should come in pairs, I propose `llvm_bools: Option<u32, u32>` (with similar updates elsewhere where both start & stop are defined & needed.

It took re-reviewing this a few times to realize that the macro define_table_getter was handling the case where one symbol was set and not the other with `self.$start?` and `self.$stop?`.

Reply:
> should come in pairs

They _should_, but we only get to see them one at a time, with an undefined order of symbol visitation, so we can't use the type you suggest.

> It took re-viewing a few times

That's fair, the macro (and much else) needs doc commenting. FYI, I wrote the various getters _without_ a macro at first, but it really was more error-prone and less clear in the end.

Reply:
Added many comments.